core/arch/arm/plat-mediatek/main.c

```diff
@@ -0,0 +1,571 @@
+extern uint8_t __heap1_end[];
+extern uint8_t __heap2_start[];
+extern uint8_t __heap2_end[];
+extern uint8_t __pageable_part_start[];
+extern uint8_t __pageable_part_end[];
+extern uint8_t __pageable_start[];
+extern uint8_t __pageable_end[];
+
+static void main_fiq(void);
+#if defined(CFG_WITH_ARM_TRUSTED_FW)
+/* Implemented in assembly, referenced in this file only */
+uint32_t cpu_on_handler(uint32_t a0, uint32_t a1);
+
+static uint32_t main_cpu_off_handler(uint32_t a0, uint32_t a1);
+static uint32_t main_cpu_suspend_handler(uint32_t a0, uint32_t a1);
+static uint32_t main_cpu_resume_handler(uint32_t a0, uint32_t a1);
+static uint32_t main_system_off_handler(uint32_t a0, uint32_t a1);
+static uint32_t main_system_reset_handler(uint32_t a0, uint32_t a1);
+#elif defined(CFG_WITH_SEC_MON)
+static uint32_t main_default_pm_handler(uint32_t a0, uint32_t a1);
+#else
+#error Platform must use either ARM_TRUSTED_FW or SEC_MON
+#endif
+
+static const struct thread_handlers handlers = {
+	.std_smc = plat_tee_entry,
+	.fast_smc = plat_tee_entry,
+	.fiq = main_fiq,
+	.svc = tee_svc_handler,
+	.abort = tee_pager_abort_handler,
+#if defined(CFG_WITH_ARM_TRUSTED_FW)
+	.cpu_on = cpu_on_handler,
+	.cpu_off = main_cpu_off_handler,
+	.cpu_suspend = main_cpu_suspend_handler,
+	.cpu_resume = main_cpu_resume_handler,
+	.system_off = main_system_off_handler,
+	.system_reset = main_system_reset_handler,
+#elif defined(CFG_WITH_SEC_MON)
+	.cpu_on = main_default_pm_handler,
+	.cpu_off = main_default_pm_handler,
+	.cpu_suspend = main_default_pm_handler,
+	.cpu_resume = main_default_pm_handler,
+	.system_off = main_default_pm_handler,
+	.system_reset = main_default_pm_handler,
+#endif
+};
+
+#if defined(CFG_WITH_ARM_TRUSTED_FW)
+static void main_init_sec_mon(uint32_t nsec_entry __unused)
+{
+	assert(nsec_entry == PADDR_INVALID);
+	/* Do nothing as we don't have a secure monitor */
+}
+#elif defined(CFG_WITH_SEC_MON)
+static void main_init_sec_mon(uint32_t nsec_entry)
+{
+	struct sm_nsec_ctx *nsec_ctx;
+
+	assert(nsec_entry != PADDR_INVALID);
+
+	/* Initialize secure monitor */
+	nsec_ctx = sm_get_nsec_ctx();
+	nsec_ctx->mon_lr = nsec_entry;
+	nsec_ctx->mon_spsr = CPSR_MODE_SVC | CPSR_I;
+
+}
+#endif
+
+#if defined(CFG_WITH_ARM_TRUSTED_FW)
+static void main_init_nsacr(void)
+{
+}
+#else
+static void main_init_nsacr(void)
+{
+	/* Normal world can use CP10 and CP11 (SIMD/VFP) */
+	write_nsacr(read_nsacr() | NSACR_CP10 | NSACR_CP11);
+}
+#endif
+
+#ifdef CFG_WITH_VFP
+static void main_init_cpacr(void)
+{
+	uint32_t cpacr = read_cpacr();
+
+	/* Enabled usage of CP10 and CP11 (SIMD/VFP) */
+	cpacr &= ~CPACR_CP(10, CPACR_CP_ACCESS_FULL);
+	cpacr |= CPACR_CP(10, CPACR_CP_ACCESS_PL1_ONLY);
+	cpacr &= ~CPACR_CP(11, CPACR_CP_ACCESS_FULL);
+	cpacr |= CPACR_CP(11, CPACR_CP_ACCESS_PL1_ONLY);
+	write_cpacr(cpacr);
+}
+#else
+static void main_init_cpacr(void)
+{
+	/* We're not using VFP/SIMD instructions, leave it disabled */
+}
+#endif
+
+#if PLATFORM_FLAVOR_IS(mt8173)
```

Comment:
Why do we have this flag in a MediaTek file only? Also I cannot see that it is working when this hasn't been set?